tests/director/tests/bundle_instance_auth_per_scenario_test.go

```diff
@@ -0,0 +1,290 @@
+	appLabel := graphql.Label{}
+	err := testctx.Tc.RunOperation(ctx, dexGraphQLClient, setAppLabelRequest, &appLabel)
+	require.NoError(t, err)
+
+	// Set runtime label
+	setRuntimeLabelRequest := fixtures.FixSetRuntimeLabelRequest(data.runtimeId, labelSelector.Key, labelSelector.Value)
+	runtimeLabel := graphql.Label{}
+	err = testctx.Tc.RunOperation(ctx, dexGraphQLClient, setRuntimeLabelRequest, &runtimeLabel)
+	require.NoError(t, err)
+
+	// Create automatic scenario assignment
+	asaInput := graphql.AutomaticScenarioAssignmentSetInput{
+		ScenarioName: otherScenario,
+		Selector:     &labelSelector,
+	}
+	fixtures.CreateAutomaticScenarioAssignmentInTenant(t, ctx, dexGraphQLClient, asaInput, tenantID)
+
+	// Cannot delete automatic scenario assignment when there are existing bundle instance auths
+	assignment := graphql.AutomaticScenarioAssignment{}
+	req := fixtures.FixDeleteAutomaticScenarioAssignmentForScenarioRequest(otherScenario)
+	err = testctx.Tc.RunOperationWithCustomTenant(ctx, dexGraphQLClient, tenantID, req, &assignment)
+	assert.Error(t, err)
+
+	// Cannot delete label which corresponds for the automatic scenario assignment
+	request := fixtures.FixDeleteRuntimeLabelRequest(data.runtimeId, labelSelector.Key)
+	label := graphql.Label{}
+	err = testctx.Tc.RunOperation(ctx, dexGraphQLClient, request, &label)
+	assert.Error(t, err)
+
+	// Deleting scenario for which there is automatic scenario assignment is successful
+	updateInput := graphql.RuntimeInput{
+		Name:   data.runtimeName,
+		Labels: graphql.Labels{ScenariosLabel: []string{commonScenario}, labelSelector.Key: labelSelector.Value},
+	}
+	updateInputGQL, err := testctx.Tc.Graphqlizer.RuntimeInputToGQL(updateInput)
+	require.NoError(t, err)
+	request = fixtures.FixUpdateRuntimeRequest(data.runtimeId, updateInputGQL)
+	input := graphql.RuntimeExt{}
+	err = testctx.Tc.RunOperation(ctx, dexGraphQLClient, request, &input)
+	require.NoError(t, err)
+
+	// Deleting bundle instance auth
+	fixtures.DeleteBundleInstanceAuth(t, ctx, dexGraphQLClient, tenantID, data.bundleInstanceAuthID)
+
+	// Now automatic scenario assignment deletion should be successful
+	req = fixtures.FixDeleteAutomaticScenarioAssignmentForScenarioRequest(otherScenario)
+	err = testctx.Tc.RunOperationWithCustomTenant(ctx, dexGraphQLClient, tenantID, req, &assignment)
+	assert.NoError(t, err)
+}
+
+func createScenarioLabelDefIfNotExist(t *testing.T, ctx context.Context) {
+	defs, err := fixtures.ListLabelDefinitionsWithinTenant(t, ctx, dexGraphQLClient, tenant.TestTenants.GetDefaultTenantID())
+	require.NoError(t, err)
+
+	for _, def := range defs {
+		if def.Key == ScenariosLabel {
+			return
+		}
+	}
+
+	fixtures.CreateScenariosLabelDefinitionWithinTenant(t, ctx, dexGraphQLClient, tenant.TestTenants.GetDefaultTenantID(), []string{conf.DefaultScenario})
+}
+
+func setupAppAndRuntimeWithCommonScenario(t *testing.T, ctx context.Context, scenario string) (*data, func(), bool) {
```

Comment:
Yeah, this is nice idea. However, I cannot avoid delegating any cleanup logic to the caller because if everything is ok the cleanup should be executed after all test cases are done